### DISCUSSION TITLE
Upgrade map tools

### DISCUSSION THREAD

**Discussion by mokun** [mars-sim/mars-sim#441] [Ideas]
Currently, we have two map tools, namely, `Mars Navigator` and `Settlement Map` in mars-sim.

`Mars Navigator` uses a large Int[] double array and stay in the memory. It is also loading inefficiently. It provides no zooming ability.

![image](https://user-images.githubusercontent.com/1168584/117550543-d62c2e00-aff5-11eb-8248-502936d0bcc9.png)

`Settlement Map` has a limited form of zooming. 

It randomly picks and chooses one of the background png icons to render a few different look of the surrounding Martian soils around the buildings of a settlement. 

![image](https://user-images.githubusercontent.com/1168584/117550580-fb20a100-aff5-11eb-812b-5bee068f1650.png)

However, it creates visual artifacts and is too homogenous in how it looks. It doesn't allow specific topological features to be added.

We should explore 

A. how to provide a higher resolution of the Martian environment by reworking our existing map tools

B. how to make it more efficient in loading map elements by regions, etc. 

C. how to allow online connection to be made to download higher resolution map tiles 

D. how to add the ability to zoom in and out of a region

E. how to add semi-permanent/temporary  topological features or man-made structures to the surrounding of a settlement (or outside a settlement) in near future ? 

Say, we want to include man-made structures such as weather stations, water drilling rigs or regolith/mineral excavation pits, etc.

Note : one related project was the height/elevation conversion of map. Please see this [tracker](https://github.com/mars-sim/mars-sim/issues/225) for past discussions.

**Comment by mokun** [comment]
@bevans2000 @Urwumpe @scud1 

Please by all means comment on A to E above.

**Comment by Urwumpe** [comment]
Well, for storing the map, I would suggest looking into Quad or kd-Trees for that task. Much faster and could also allow lazy loading. Orbiter uses something similar for storing its terrain data for the solar system.

I wouldn't want to go online, since going online also requires online infrastructure to be used. Not very easily done by a small project and TANSTAAFL.

For the settlement map... honestly no idea, except using a texture behind that is continous. 

More enhancement might be storing the settlement map data more efficient and leave it to the client to translate it into a visual representation. No problem adding additional details to the maps (boulders, outcrops, cliffs) to make them look more interesting, a voxel model of the local aerology could also be nice, if you would like to go into the third dimension.

But that all is a lot of work for little actual improvement.

**Reply by mokun** [reply]
We have recently made some positive change in bundling with better resolution of global maps and now even have provision for downloading higher resolution maps during the runtime. 

See #852 and #895.

However, we still haven't got to the point of tile loading yet.

And we'll need to deal with improving settlement maps in hopefully near future.
